pkg/platform/authentication/auth.go

```diff
@@ -215,6 +249,57 @@ func (s *Auth) AuthenticateWithToken(token string) error {
 	})
 }
 
+// AuthenticateWithDeviceCode posts a request to authenticate this device on the Platform and waits
+// for the user to authorize the request.
+// The given callback function is called when the user should be prompted to authorize the request.
+// Returns an error if authentication cannot be performed (e.g. timeout or Platform is unreachable).
+func (s *Auth) AuthenticateWithDevice(promptCallback func(userCode, uri string)) error {
+	if s.Authenticated() {
+		return nil // nothing to do
+	}
+	// Post the authentication request to the Platform.
+	postParams := oauth.NewAuthDevicePostParams()
+	response, err := mono.Get().Oauth.AuthDevicePost(postParams)
+	if err != nil {
+		logging.Error("Error requesting device authentication: %v", err)
+		return locale.NewError("err_auth_device")
+	}
+	// Prompt the user to authorize the request.
+	promptCallback(*response.Payload.UserCode, *response.Payload.VerificationURIComplete)
+	// Wait for the authorization.
+	deviceCode := strfmt.UUID(*response.Payload.DeviceCode)
+	getParams := oauth.NewAuthDeviceGetParams()
+	getParams.SetDeviceCode(deviceCode)
+	startTime := time.Now()
+	const timeout = 6 * 60 * time.Second
+	for {
+		response, err := mono.Get().Oauth.AuthDeviceGet(getParams)
+		if response != nil {
+			s.storeAuthenticatedDevice(deviceCode, response)
+			break
+		} else if errs.Matches(err, &oauth.AuthDeviceGetBadRequest{}) {
+			badRequest := err.(*oauth.AuthDeviceGetBadRequest)
+			errorString := *badRequest.Payload.Error
+			const expiredToken = "expired_token"
+			const invalidClient = "invalid_client"
+			const slowDown = "slow_down"
+			if errorString == expiredToken || time.Since(startTime) >= timeout {
+				return locale.NewInputError("auth_device_timeout")
+			} else if errorString == invalidClient {
```

Comment:
We want to avoid string matching like this, unless it's done via swagger generated types. Is this errorString being defined in the swagger code? If so re-generating the mono-api should give you these strings: http://github.com/ActiveState/cli/blob/b94e2af417de13db488fc241fcfa86c24f8ede6e/activestate.generators.yaml#L2-L2

Reply:
Great, that's just what I was looking for.